python/ray/data/dataset_pipeline.py

```diff
@@ -497,7 +499,9 @@ def schema(
             The Python type or Arrow schema of the records, or None if the
             schema is not known.
         """
-        return self._peek().schema(fetch_if_missing=fetch_if_missing)
+        if self._executed[0]:
+            raise RuntimeError("Pipeline cannot read schema after execution.")
```

Comment:
Another option is we could always save it when the first dataset is available, not sure if that's easy to implement.

Reply:
It's easy. I made it so now.